File: python/tests/searchoptions_test.py
# -*- coding: utf-8 -*-
################################################################################
#
# searchoptions_test.py
#
# class SearchOptionsTest: testing of SearchOptions class
#
################################################################################
import re
import sys
import unittest

sys.path.insert(1, '../pysearch')

from searchexception import SearchException
from searchoptions import SearchOptions
from searchsettings import SearchSettings


class SearchOptionsTest(unittest.TestCase):
    @classmethod
    def setUpClass(self):
        self.searchoptions = SearchOptions()

    def test_no_args(self):
        # test the props
        settings = self.searchoptions.search_settings_from_args([])
        self.assertFalse(settings.archivesonly)
        self.assertFalse(settings.debug)
        self.assertFalse(settings.firstmatch)
        self.assertTrue(settings.excludehidden)
        self.assertEqual(settings.linesafter, 0)
        self.assertEqual(settings.linesbefore, 0)
        self.assertFalse(settings.listdirs)
        self.assertFalse(settings.listfiles)
        self.assertFalse(settings.listlines)
        self.assertEqual(settings.maxlinelength, 150)
        self.assertFalse(settings.multilinesearch)
        self.assertTrue(settings.printresults)
        self.assertFalse(settings.printusage)
        self.assertFalse(settings.printversion)
        self.assertTrue(settings.recursive)
        self.assertFalse(settings.searcharchives)
        self.assertFalse(settings.uniquelines)
        self.assertFalse(settings.verbose)
        # test the extensino and pattern sets
        self.assertFalse(settings.in_archiveextensions)
        self.assertFalse(settings.in_archivefilepatterns)
        self.assertFalse(settings.in_dirpatterns)
        self.assertFalse(settings.in_filepatterns)
        self.assertFalse(settings.in_linesafterpatterns)
        self.assertFalse(settings.in_linesbeforepatterns)
        self.assertFalse(settings.linesaftertopatterns)
        self.assertFalse(settings.linesafteruntilpatterns)
        self.assertFalse(settings.out_archiveextensions)
        self.assertFalse(settings.out_archivefilepatterns)
        self.assertFalse(settings.out_dirpatterns)
        self.assertFalse(settings.out_filepatterns)
        self.assertFalse(settings.out_linesafterpatterns)
        self.assertFalse(settings.out_linesbeforepatterns)
        self.assertFalse(settings.searchpatterns)

    def test_valid_args(self):
        args = ['-x', 'py,rb', '-s', 'Search', '.']
        settings = self.searchoptions.search_settings_from_args(args)
        self.assertEqual(settings.startpath, '.')
        for x in set(['py', 'rb']):
            self.assertIn(x, settings.in_extensions)
        self.assertEqual(list(settings.searchpatterns)[0].pattern, 'Search')

    def test_archivesonly_arg(self):
        args = ['--archivesonly']
        settings = self.searchoptions.search_settings_from_args(args)
        self.assertTrue(settings.archivesonly)
        self.assertTrue(settings.searcharchives)

    def test_debug_arg(self):
        args = ['--debug']
        settings = self.searchoptions.search_settings_from_args(args)
        self.assertTrue(settings.debug)
        self.assertTrue(settings.verbose)

    def test_missing_arg(self):
        args = ['-x', 'py,rb', '-s', 'Search', '.', '-D']
        with self.assertRaises(SearchException) as cm:
            settings = self.searchoptions.search_settings_from_args(args)
        self.assertEqual(str(cm.exception), 'Missing value for option D')

    def test_invalid_arg(self):
        args = ['-x', 'py,rb', '-s', 'Search', '.', '-Q']
        with self.assertRaises(SearchException) as cm:
            settings = self.searchoptions.search_settings_from_args(args)
        self.assertEqual(str(cm.exception), 'Invalid option: Q')


    def test_settings_from_json(self):
        settings = SearchSettings()
        json = '''{
  "startpath": "~/src/xsearch/",
  "in-ext": ["js","ts"],
  "out-dirpattern": "node_module",
  "out-filepattern": ["temp"],
  "search": "Searcher",
  "linesbefore": 2,
  "linesafter": 2,
  "debug": true,
  "allmatches": false,
  "includehidden": true
}'''
        self.searchoptions.settings_from_json(json, settings)
        self.assertEqual(settings.startpath, '~/src/xsearch/')
        for x in set(['js', 'ts']):
            self.assertIn(x, settings.in_extensions)
        self.assertEqual(list(settings.searchpatterns)[0].pattern, 'Searcher')
        self.assertEqual(list(settings.out_dirpatterns)[0].pattern, 'node_module')
        self.assertEqual(list(settings.out_filepatterns)[0].pattern, 'temp')
        self.assertEqual(settings.linesbefore, 2)
        self.assertEqual(settings.linesafter, 2)
        self.assertTrue(settings.debug)
        self.assertTrue(settings.verbose)
        self.assertTrue(settings.firstmatch)
        self.assertFalse(settings.excludehidden)


if __name__ == '__main__':
    unittest.main()

File: ruby/tests/searchoptions_test.rb
################################################################################
#
# searchoptions_test.rb
#
# Test the SearchOptions class
#
################################################################################

require_relative "../rbsearch/searchoptions.rb"
require "test/unit"
 
class SearchOptionsTest < Test::Unit::TestCase
  def setup
    @searchoptions = SearchOptions.new
  end

  def test_no_args
    settings = @searchoptions.search_settings_from_args([])
    assert_equal(settings.archivesonly, false)
    assert_equal(settings.debug, false)
    assert_equal(settings.excludehidden, true)
    assert_equal(settings.firstmatch, false)
    assert_equal(settings.linesafter, 0)
    assert_equal(settings.linesbefore, 0)
    assert_equal(settings.listdirs, false)
    assert_equal(settings.listfiles, false)
    assert_equal(settings.listlines, false)
    assert_equal(settings.maxlinelength, 150)
    assert_equal(settings.multilinesearch, false)
    assert_equal(settings.printresults, true)
    assert_equal(settings.printusage, false)
    assert_equal(settings.printversion, false)
    assert_equal(settings.recursive, true)
    assert_equal(settings.searcharchives, false)
    assert_equal(settings.startpath, nil)
    assert_equal(settings.uniquelines, false)
    assert_equal(settings.verbose, false)
    assert(settings.in_archiveextensions.empty?)
    assert(settings.in_archivefilepatterns.empty?)
    assert(settings.in_dirpatterns.empty?)
    assert(settings.in_filepatterns.empty?)
    assert(settings.in_linesafterpatterns.empty?)
    assert(settings.in_linesbeforepatterns.empty?)
    assert(settings.linesaftertopatterns.empty?)
    assert(settings.linesafteruntilpatterns.empty?)
    assert(settings.out_archiveextensions.empty?)
    assert(settings.out_archivefilepatterns.empty?)
    assert(settings.out_dirpatterns.empty?)
    assert(settings.out_filepatterns.empty?)
    assert(settings.out_linesafterpatterns.empty?)
    assert(settings.out_linesbeforepatterns.empty?)
    assert(settings.searchpatterns.empty?)
  end

  def test_valid_args
    args = ['-x', 'py,rb', '-s', 'Search', '.']
    settings = @searchoptions.search_settings_from_args(args)
    assert_equal(settings.startpath, '.')
    assert_equal(settings.in_extensions.length, 2)
    assert(settings.in_extensions.include?('py'))
    assert(settings.in_extensions.include?('rb'))
    assert_equal(settings.searchpatterns.length, 1)
    assert_equal(settings.searchpatterns.first.source, 'Search')
  end

  def test_archivesonly_arg
    args = ['--archivesonly']
    settings = @searchoptions.search_settings_from_args(args)
    assert(settings.archivesonly)
    assert(settings.searcharchives)
  end

  def test_debug_arg
    args = ['--debug']
    settings = @searchoptions.search_settings_from_args(args)
    assert(settings.debug)
    assert(settings.verbose)
  end

  def test_with_invalid_arg
    args = ['-x', 'py,rb', '-s', 'Search', '.', '-Q']
    assert_raise(ArgumentError) {settings = @searchoptions.search_settings_from_args(args)}
  end
end
